apps/Gruntfile.js

```diff
@@ -706,22 +706,99 @@ describe('entry tests', () => {
+              chunks: chunk => {
+                return _.keys(appsEntries).includes(chunk.name);
+              }
+            },
+            // Pull any module shared by 2+ codeStudioEntries into the
+            // "code-studio-common" chunk.
+            'code-studio-common': {
+              name: 'code-studio-common',
+              minChunks: 2,
+              chunks: chunk => {
+                const chunkNames = _.keys(codeStudioEntries);
+                return chunkNames.includes(chunk.name);
+              },
+              priority: 10
+            },
+            // With just the cacheGroups listed above, we end up with many
+            // duplicate modules between the "common" and "code-studio-common"
+            // chunks. The next cache group eliminates some of this duplication
+            // by pulling more modules from "common" into "code-studio-common".
+            //
+            // The use of minChunks provides a guarantee that we don't
+            // unnecessarily move things into "code-studio-common" which are
+            // needed only by appsEntries. This avoids increasing the download
+            // size for code studio pages which include code-studio-common.js
+            // but not common.js.
+            //
+            // There is no converse guarantee that this strategy will eliminate
+            // all duplication between "common" and "code-studio-common".
+            // However, at the time of this writing, bundle analysis indicates
+            // that is currently effective in eliminating any duplication.
+            //
+            // In the future, we want to move toward asynchronous imports, which
+            // allow webpack to manage bundle splitting and sharing behind the
+            // scenes. Once we adopt this approach, the need for predefined
+            // cacheGroups will go away.
+            //
+            // For more information see: https://webpack.js.org/guides/code-splitting/
+            'code-studio-multi': {
+              name: 'code-studio-common',
+              minChunks: _.keys(appsEntries).length + 1,
+              chunks: chunk => {
+                const chunkNames = _.keys(codeStudioEntries).concat(
+                  _.keys(appsEntries)
+                );
+                return chunkNames.includes(chunk.name);
+              },
+              priority: 20
+            }
+          }
+        }
+      },
       plugins: [
-        new webpack.optimize.CommonsChunkPlugin({
-          name: 'common',
-          chunks: _.keys(appsEntries),
-          minChunks: 2
-        }),
-        new webpack.optimize.CommonsChunkPlugin({
-          name: 'code-studio-common',
-          chunks: _.keys(codeStudioEntries).concat(['common']),
-          minChunks: 2
-        }),
-        new webpack.optimize.CommonsChunkPlugin({
-          name: 'essential',
-          minChunks: 2,
-          chunks: ['peer_reviews', 'plc', 'pd', 'code-studio-common']
```

Comment:
The small regression in chunk sizes in 'plc' and 'code-studio-common' is due to this "essential" cache group going away. 

'pd' and 'peer_reviews' were bogus entries that did not match any entry points in our system.

Reply:
Worth pulling some of this cleanup into a pre-flight PR? If it's not tied to the Webpack 4 upgrade directly.

Reply:
@joshlory I like the idea of a pre-flight PR, and if we are going to pre-flight anything, it would probably be this removal of the 'essential' chunk config which you've identified. But, I'm having trouble convincing myself that this eliminates enough complexity to warrant a pre-flight PR, especially considering that we've done a reasonable amount of analysis here to convince ourselves that any user-facing performance problems will be tolerable. 

There is no immediate rush here though, so if you think this would help reduce the risk of problems I'd be fine with splitting this block into it's own PR.

Reply:
No strong preference, whatever you think is best!

Preflighting has been useful in some of the bigger Rails upgrades, but this upgrade seems pretty safe 🤞.